File: templates/Quick_Sort.cpp
#include <iostream>
#include <stack>
using namespace std;

// 把数组分为两部分，轴temp左边的部分都小于轴右边的部分
int partition(int* nums, int low, int high)
{
	int temp = nums[low];		//任选元素作为轴，这里选首元素
	while (low < high)
	{
		while (low < high && nums[high] >= temp) high--;
		nums[low] = nums[high];
		while (low < high && nums[low] <= temp) low++;
		nums[high] = nums[low];
	}
	//此时low == high
	nums[low] = temp;
	return low;
}

// 递归
void quickSort(int* nums, int low, int high)
{
	if (low < high)
	{
		int mid = partition(nums, low, high);
		quickSort(nums, low, mid - 1);
		quickSort(nums, mid + 1, high);
	}
}

// 非递归（利用栈）
// 其实就是用栈保存每一个待排序子串的首尾元素下标，
// 下一次while循环时取出这个范围，对这段子序列进行partition操作
void quickSort(int* nums, int low, int high)
{
	if (low >= high) return;
	stack<int> sta;
	sta.push(low);
	sta.push(high);
	while (!sta.empty())
	{
		int j = sta.top();
		sta.pop();
		int i = sta.top();
		sta.pop();
		int mid = partition(nums, i, j);
		if (i < mid - 1)
		{
			sta.push(i);
			sta.push(mid - 1);
		}
		if (mid + 1 < j)
		{
			sta.push(mid + 1);
			sta.push(j);
		}
	}
}

int main()
{
	int nums[] = { 4, 2, 3, 9, 5, 6, 8, 7, 1, 0 };
	quickSort(nums, 0, 9);

	for (int i = 0; i < 10; ++i)
		cout << nums[i] << " ";
	cout << endl;

	system("pause");
	return 0;
}

File: templates/Quick_Sort.java
/**
* 本参考程序来自九章算法，由 @马克助教 提供。版权所有，转发请注明出处。
* - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
* - 现有的面试培训课程包括：九章算法班，系统设计班，算法强化班，Java入门与基础算法班，Android 项目实战班，
* - Big Data 项目实战班，算法面试高频题班, 动态规划专题班
* - 更多详情请见官方网站：http://www.jiuzhang.com/?source=code
*/

import java.util.Random;

public class Solution {
    /*
     * @param A: an integer array
     * @return:
     */
    public Random rand;
    public void sortIntegers2(int[] A) {
        rand = new Random();
        // write your code here
        quickSort(A, 0, A.length - 1);
    }

    public void quickSort(int[] A, int start, int end) {
        if (start >= end) {
            return;
        }

        int index = rand.nextInt(end - start + 1)  + start;
        int pivot = A[index];
        int left = start;
        int right = end;

        while (left <= right) {
            while (left <= right && A[left] < pivot) {
                left ++;
            }
            while (left <= right && A[right] > pivot) {
                right --;
            }

            if (left <= right) {
                int temp = A[left];
                A[left] = A[right];
                A[right] = temp;

                left ++;
                right --;
            }
        }
        // A[start... right]
        quickSort(A, start, right);
        // A[left ... end]
        quickSort(A, left, end);
    }
}







/** Java program for implementation of QuickSort */
class QuickSort
{
    /* This function takes last element as pivot,
       places the pivot element at its correct
       position in sorted array, and places all
       smaller (smaller than pivot) to left of
       pivot and all greater elements to right
       of pivot */
    int partition(int arr[], int low, int high)
    {
        int pivot = arr[high];
        int i = (low - 1); // index of smaller element
        for (int j = low; j < high; j++)
        {
            // If current element is smaller than or
            // equal to pivot
            if (arr[j] <= pivot)
            {
                i++;

                // swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // swap arr[i+1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    /* The main function that implements QuickSort()
      arr[] --> Array to be sorted,
      low   --> Starting index,
      high  --> Ending index */
    void sort(int arr[], int low, int high)
    {
        if (low < high)
        {
            /* pi is partitioning index, arr[pi] is
              now at right place */
            int pi = partition(arr, low, high);

            // Recursively sort elements before
            // partition and after partition
            sort(arr, low, pi - 1);
            sort(arr, pi + 1, high);
        }
    }

    /* A utility function to print array of size n */
    static void printArray(int arr[])
    {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    // Driver program
    public static void main(String args[])
    {
        int arr[] = {10, 7, 8, 9, 1, 5};
        int n = arr.length;

        QuickSort ob = new QuickSort();
        ob.sort(arr, 0, n - 1);

        System.out.println("sorted array");
        printArray(arr);
    }
}
